kernel/thread_abort.c

```diff
@@ -21,25 +21,48 @@
 #include <ksched.h>
 #include <sys/__assert.h>
 #include <syscall_handler.h>
+#include <logging/log.h>
+LOG_MODULE_DECLARE(os);
+
+FUNC_NORETURN void z_self_abort(void)
+{
+	/* Self-aborting threads don't clean themselves up, we
+	 * have the idle thread for the current CPU do it.
+	 */
+	int key;
+	struct _cpu *cpu;
+
+	/* Lock local IRQs to prevent us from migrating to another CPU
+	 * while we set this up
+	 */
+	key = arch_irq_lock();
+	cpu = _current_cpu;
+	cpu->pending_abort = _current;
```

Comment:
shouldn't this be a list to handle multiple threads self-aborting before we are able to switch to idle to cleanup?

Reply:
can you give some detail how that could happen?

I think if a thread self-aborts, the next thing that gets scheduled is necessarily the idle thread on the same CPU that the thread aborted on. I don't see how multiple could queue up since it should impossible for another thread to run in between on this CPU.

Reply:
Just noticed the change in next_up() which will put the idle thread next if there is something to abort, so you are correct.

Before:

```
1969-12-31 19:00:00.005629 (+0.000148 s): thread_pending: thread_b
1969-12-31 19:00:00.005778 (+0.000150 s): thread_switched_out: thread_b
1969-12-31 19:00:00.005829 (+0.000050 s): thread_switched_in: thread_c
1969-12-31 19:00:00.005887 (+0.000058 s): thread_abort: thread_c
1969-12-31 19:00:00.005937 (+0.000050 s): thread_switched_out: thread_c
1969-12-31 19:00:00.005985 (+0.000048 s): thread_switched_in: thread_d
1969-12-31 19:00:00.006050 (+0.000065 s): thread_abort: thread_d
1969-12-31 19:00:00.006110 (+0.000061 s): thread_switched_out: thread_d
1969-12-31 19:00:00.006159 (+0.000049 s): thread_switched_in: idle 00
1969-12-31 19:00:00.006295 (+0.000135 s): idle
```
2 threads are abort, then we go into idle..

Now
```
1969-12-31 19:00:00.005027 (+0.000041 s): thread_switched_in: thread_c
1969-12-31 19:00:00.005075 (+0.000048 s): thread_suspend: thread_c
1969-12-31 19:00:00.005116 (+0.000041 s): thread_switched_out: thread_c
1969-12-31 19:00:00.005156 (+0.000040 s): thread_switched_in: idle 00
1969-12-31 19:00:00.005222 (+0.000066 s): thread_abort: thread_c
1969-12-31 19:00:00.005273 (+0.000051 s): thread_switched_out: idle 00
1969-12-31 19:00:00.005313 (+0.000040 s): thread_switched_in: thread_d
1969-12-31 19:00:00.005364 (+0.000051 s): thread_suspend: thread_d
1969-12-31 19:00:00.005405 (+0.000041 s): thread_switched_out: thread_d
1969-12-31 19:00:00.005444 (+0.000039 s): thread_switched_in: idle 00
1969-12-31 19:00:00.005485 (+0.000041 s): thread_abort: thread_d
1969-12-31 19:00:00.005536 (+0.000051 s): thread_switched_out: idle 00
1969-12-31 19:00:00.005577 (+0.000041 s): thread_switched_in: idle 00
```